README.md
<div align="center">

# captcha.gif

**A simple captcha generator for nodejs.**

[![npm](https://img.shields.io/npm/dw/captcha.gif)](https://www.npmjs.com/package/captcha.gif)
[![codecov](https://codecov.io/gh/imranbarbhuiya/captcha.gif/branch/main/graph/badge.svg)](https://codecov.io/gh/imranbarbhuiya/captcha.gif)
[![npm](https://img.shields.io/npm/v/captcha.gif?color=crimson&logo=npm&style=flat-square)](https://www.npmjs.com/package/captcha.gif)

</div>

## Description

A simple lightweight captcha generator for nodejs.

Impliments [captcha generator in C](http://brokestream.com/captcha.html) in js.

Read TypeDoc generated Documentation [here](https://imranbarbhuiya.github.io/captcha.gif/).

## Features

-   Written In Typescript
-   Offers CJS, ESM builds
-   Full TypeScript & JavaScript support

## Install

```bash
# npm
npm i captcha.gif

# yarn
yarn add captcha.gif

```

## Usage

```ts
import { Captcha } from 'captcha.gif';
```

```js
const { Captcha } = require('captcha.gif');
```

### Generate Captcha

```ts
const captcha = new Captcha();

const { token, buffer } = captcha.generate();
```

## Buy me some doughnuts

If you want to support me by donating, you can do so by using any of the following methods. Thank you very much in advance!

<a href="https://www.buymeacoffee.com/parbez" target="_blank"><img src="https://cdn.buymeacoffee.com/buttons/default-orange.png" alt="Buy Me A Coffee" height="41" width="174"></a>
<a href='https://ko-fi.com/Y8Y1CBIJH' target='_blank'><img height='36' style='border:0px;height:36px;' src='https://cdn.ko-fi.com/cdn/kofi4.png?v=3' border='0' alt='Buy Me a Coffee at ko-fi.com' /></a>

## Contributors ✨

Thanks goes to these wonderful people:

<a href="https://github.com/imranbarbhuiya/TagScript/graphs/contributors">
    <img src="https://contrib.rocks/image?repo=imranbarbhuiya/TagScript" />
</a>


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

# [1.0.11](https://github.com/imranbarbhuiya/captcha.gif/compare/v1.0.10...v1.0.11) - (2023-06-18)

## 🐛 Bug Fixes

- Remove all contrib ([588ed2d](https://github.com/imranbarbhuiya/captcha.gif/commit/588ed2dee41085dd7f388879a62ef06f069fd320))

# [1.0.9](https://github.com/imranbarbhuiya/captcha.gif/compare/v1.0.8...v1.0.9) - (2022-09-13)

## 🐛 Bug Fixes

- Readd dts ([024cddf](https://github.com/imranbarbhuiya/captcha.gif/commit/024cddfa71ee8a83f2589a38325219577505056f))



tests/Captcha.test.ts
import { Captcha } from '../src';

describe('Captcha', () => {
	const captcha = new Captcha();

	test('generate should return a token and a buffer', () => {
		const { token, buffer } = captcha.generate(4);

		expect(token).toBeDefined();
		expect<string>(typeof token).toBe('string');
		expect<Buffer>(buffer).toBeInstanceOf(Buffer);
	});

	test('generate buffer with length 17646', () => {
		const { buffer } = captcha.generate(4);

		expect(buffer.length).toBe(17646);
	});

	test('GIVEN no size THEN return length 5 token', () => {
		const { token } = captcha.generate();

		expect<string>(typeof token).toBe('string');
		expect(token.length).toBe(5);
	});

	test('GIVEN size THEN return length size token', () => {
		const { token } = captcha.generate(10);

		expect<string>(typeof token).toBe('string');
		expect(token.length).toBe(10);
	});
});


src/lib/Sw.ts
export const SW = [
	0, 4, 8, 12, 16, 20, 23, 27, 31, 35, 39, 43, 47, 50, 54, 58, 61, 65, 68, 71, 75, 78, 81, 84, 87, 90, 93, 96, 98, 101, 103, 105, 108, 110, 112,
	114, 115, 117, 119, 120, 121, 122, 123, 124, 125, 126, 126, 127, 127, 127, 127, 127, 127, 127, 126, 126, 125, 124, 123, 122, 121, 120, 119, 117,
	115, 114, 112, 110, 108, 105, 103, 101, 98, 96, 93, 90, 87, 84, 81, 78, 75, 71, 68, 65, 61, 58, 54, 50, 47, 43, 39, 35, 31, 27, 23, 20, 16, 12, 8,
	4, 0, -4, -8, -12, -16, -20, -23, -27, -31, -35, -39, -43, -47, -50, -54, -58, -61, -65, -68, -71, -75, -78, -81, -84, -87, -90, -93, -96, -98,
	-101, -103, -105, -108, -110, -112, -114, -115, -117, -119, -120, -121, -122, -123, -124, -125, -126, -126, -127, -127, -127, -127, -127, -127,
	-127, -126, -126, -125, -124, -123, -122, -121, -120, -119, -117, -115, -114, -112, -110, -108, -105, -103, -101, -98, -96, -93, -90, -87, -84,
	-81, -78, -75, -71, -68, -65, -61, -58, -54, -50, -47, -43, -39, -35, -31, -27, -23, -20, -16, -12, -8, -4
];


assets/search.js
window.searchData = JSON.parse("{\"kinds\":{\"128\":\"Class\",\"256\":\"Interface\",\"512\":\"Constructor\",\"1024\":\"Property\",\"2048\":\"Method\",\"4194304\":\"Type alias\"},\"rows\":[{\"id\":0,\"kind\":128,\"name\":\"Captcha\",\"url\":\"classes/Captcha.html\",\"classes\":\"tsd-kind-class\"},{\"id\":1,\"kind\":512,\"name\":\"constructor\",\"url\":\"classes/Captcha.html#constructor\",\"classes\":\"tsd-kind-constructor tsd-parent-kind-class\",\"parent\":\"Captcha\"},{\"id\":2,\"kind\":1024,\"name\":\"numberOfDots\",\"url\":\"classes/Captcha.html#numberOfDots\",\"classes\":\"tsd-kind-property tsd-parent-kind-class tsd-is-private\",\"parent\":\"Captcha\"},{\"id\":3,\"kind\":1024,\"name\":\"gifSize\",\"url\":\"classes/Captcha.html#gifSize\",\"classes\":\"tsd-kind-property tsd-parent-kind-class tsd-is-private\",\"parent\":\"Captcha\"},{\"id\":4,\"kind\":1024,\"name\":\"letters\",\"url\":\"classes/Captcha.html#letters\",\"classes\":\"tsd-kind-property tsd-parent-kind-class tsd-is-private\",\"parent\":\"Captcha\"},{\"id\":5,\"kind\":1024,\"name\":\"blurImg\",\"url\":\"classes/Captcha.html#blurImg\",\"classes\":\"tsd-kind-property tsd-parent-kind-class tsd-is-private\",\"parent\":\"Captcha\"},{\"id\":6,\"kind\":1024,\"name\":\"filterImg\",\"url\":\"classes/Captcha.html#filterImg\",\"classes\":\"tsd-kind-property tsd-parent-kind-class tsd-is-private\",\"parent\":\"Captcha\"},{\"id\":7,\"kind\":2048,\"name\":\"generate\",\"url\":\"classes/Captcha.html#generate\",\"classes\":\"tsd-kind-method tsd-parent-kind-class\",\"parent\":\"Captcha\"},{\"id\":8,\"kind\":2048,\"name\":\"random\",\"url\":\"classes/Captcha.html#random\",\"classes\":\"tsd-kind-method tsd-parent-kind-class tsd-is-private\",\"parent\":\"Captcha\"},{\"id\":9,\"kind\":2048,\"name\":\"letter\",\"url\":\"classes/Captcha.html#letter\",\"classes\":\"tsd-kind-method tsd-parent-kind-class tsd-is-private\",\"parent\":\"Captcha\"},{\"id\":10,\"kind\":2048,\"name\":\"line\",\"url\":\"classes/Captcha.html#line\",\"classes\":\"tsd-kind-method tsd-parent-kind-class tsd-is-private\",\"parent\":\"Captcha\"},{\"id\":11,\"kind\":2048,\"name\":\"dots\",\"url\":\"classes/Captcha.html#dots\",\"classes\":\"tsd-kind-method tsd-parent-kind-class tsd-is-private\",\"parent\":\"Captcha\"},{\"id\":12,\"kind\":2048,\"name\":\"blur\",\"url\":\"classes/Captcha.html#blur\",\"classes\":\"tsd-kind-method tsd-parent-kind-class tsd-is-private\",\"parent\":\"Captcha\"},{\"id\":13,\"kind\":2048,\"name\":\"filter\",\"url\":\"classes/Captcha.html#filter\",\"classes\":\"tsd-kind-method tsd-parent-kind-class tsd-is-private\",\"parent\":\"Captcha\"},{\"id\":14,\"kind\":2048,\"name\":\"makeCaptcha\",\"url\":\"classes/Captcha.html#makeCaptcha\",\"classes\":\"tsd-kind-method tsd-parent-kind-class tsd-is-private\",\"parent\":\"Captcha\"},{\"id\":15,\"kind\":2048,\"name\":\"makeGif\",\"url\":\"classes/Captcha.html#makeGif\",\"classes\":\"tsd-kind-method tsd-parent-kind-class tsd-is-private\",\"parent\":\"Captcha\"},{\"id\":16,\"kind\":256,\"name\":\"Options\",\"url\":\"interfaces/Options.html\",\"classes\":\"tsd-kind-interface\"},{\"id\":17,\"kind\":1024,\"name\":\"numberOfDots\",\"url\":\"interfaces/Options.html#numberOfDots\",\"classes\":\"tsd-kind-property tsd-parent-kind-interface\",\"parent\":\"Options\"},{\"id\":18,\"kind\":1024,\"name\":\"blur\",\"url\":\"interfaces/Options.html#blur\",\"classes\":\"tsd-kind-property tsd-parent-kind-interface\",\"parent\":\"Options\"},{\"id\":19,\"kind\":1024,\"name\":\"filter\",\"url\":\"interfaces/Options.html#filter\",\"classes\":\"tsd-kind-property tsd-parent-kind-interface\",\"parent\":\"Options\"},{\"id\":20,\"kind\":4194304,\"name\":\"Color\",\"url\":\"modules.html#Color\",\"classes\":\"tsd-kind-type-alias\"}],\"index\":{\"version\":\"2.3.9\",\"fields\":[\"name\",\"parent\"],\"fieldVectors\":[[\"name/0\",[0,2.877]],[\"parent/0\",[]],[\"name/1\",[1,26.856]],[\"parent/1\",[0,0.269]],[\"name/2\",[2,21.748]],[\"parent/2\",[0,0.269]],[\"name/3\",[3,26.856]],[\"parent/3\",[0,0.269]],[\"name/4\",[4,26.856]],[\"parent/4\",[0,0.269]],[\"name/5\",[5,26.856]],[\"parent/5\",[0,0.269]],[\"name/6\",[6,26.856]],[\"parent/6\",[0,0.269]],[\"name/7\",[7,26.856]],[\"parent/7\",[0,0.269]],[\"name/8\",[8,26.856]],[\"parent/8\",[0,0.269]],[\"name/9\",[9,26.856]],[\"parent/9\",[0,0.269]],[\"name/10\",[10,26.856]],[\"parent/10\",[0,0.269]],[\"name/11\",[11,26.856]],[\"parent/11\",[0,0.269]],[\"name/12\",[12,21.748]],[\"parent/12\",[0,0.269]],[\"name/13\",[13,21.748]],[\"parent/13\",[0,0.269]],[\"name/14\",[14,26.856]],[\"parent/14\",[0,0.269]],[\"name/15\",[15,26.856]],[\"parent/15\",[0,0.269]],[\"name/16\",[16,15.87]],[\"parent/16\",[]],[\"name/17\",[2,21.748]],[\"parent/17\",[16,1.486]],[\"name/18\",[12,21.748]],[\"parent/18\",[16,1.486]],[\"name/19\",[13,21.748]],[\"parent/19\",[16,1.486]],[\"name/20\",[17,26.856]],[\"parent/20\",[]]],\"invertedIndex\":[[\"blur\",{\"_index\":12,\"name\":{\"12\":{},\"18\":{}},\"parent\":{}}],[\"blurimg\",{\"_index\":5,\"name\":{\"5\":{}},\"parent\":{}}],[\"captcha\",{\"_index\":0,\"name\":{\"0\":{}},\"parent\":{\"1\":{},\"2\":{},\"3\":{},\"4\":{},\"5\":{},\"6\":{},\"7\":{},\"8\":{},\"9\":{},\"10\":{},\"11\":{},\"12\":{},\"13\":{},\"14\":{},\"15\":{}}}],[\"color\",{\"_index\":17,\"name\":{\"20\":{}},\"parent\":{}}],[\"constructor\",{\"_index\":1,\"name\":{\"1\":{}},\"parent\":{}}],[\"dots\",{\"_index\":11,\"name\":{\"11\":{}},\"parent\":{}}],[\"filter\",{\"_index\":13,\"name\":{\"13\":{},\"19\":{}},\"parent\":{}}],[\"filterimg\",{\"_index\":6,\"name\":{\"6\":{}},\"parent\":{}}],[\"generate\",{\"_index\":7,\"name\":{\"7\":{}},\"parent\":{}}],[\"gifsize\",{\"_index\":3,\"name\":{\"3\":{}},\"parent\":{}}],[\"letter\",{\"_index\":9,\"name\":{\"9\":{}},\"parent\":{}}],[\"letters\",{\"_index\":4,\"name\":{\"4\":{}},\"parent\":{}}],[\"line\",{\"_index\":10,\"name\":{\"10\":{}},\"parent\":{}}],[\"makecaptcha\",{\"_index\":14,\"name\":{\"14\":{}},\"parent\":{}}],[\"makegif\",{\"_index\":15,\"name\":{\"15\":{}},\"parent\":{}}],[\"numberofdots\",{\"_index\":2,\"name\":{\"2\":{},\"17\":{}},\"parent\":{}}],[\"options\",{\"_index\":16,\"name\":{\"16\":{}},\"parent\":{\"17\":{},\"18\":{},\"19\":{}}}],[\"random\",{\"_index\":8,\"name\":{\"8\":{}},\"parent\":{}}]],\"pipeline\":[]}}");

src/lib/Colors.ts
export const colors = [
	// Black 500 #000000
	`GIF89a\xc8\0\x46\0\x83\0\0
\x11\x11\x11
\x11\x11\x11
\x11\x11\x11
\x11\x11\x11
\x11\x11\x11
\x11\x11\x11
\x11\x11\x11
\x11\x11\x11
\x11\x11\x11
\x11\x11\x11
\x11\x11\x11
\x11\x11\x11
\x11\x11\x11
\x11\x11\x11
\x11\x11\x11
\xff\xff\xff,\0\0\0\0\xc8\0\x46\0\0\x04`,

	// Red 500 #F44336
	`GIF89a\xc8\0\x46\0\x83\0\0
\xF4\x43\x36
\xF4\x43\x36
\xF4\x43\x36
\xF4\x43\x36
\xF4\x43\x36
\xF4\x43\x36
\xF4\x43\x36
\xF4\x43\x36
\xF4\x43\x36
\xF4\x43\x36
\xF4\x43\x36
\xF4\x43\x36
\xF4\x43\x36
\xF4\x43\x36
\xF4\x43\x36
\xff\xff\xff,\0\0\0\0\xc8\0\x46\0\0\x04`,

	// Pink 500 #E91E63
	`GIF89a\xc8\0\x46\0\x83\0\0
\xE9\x1E\x63
\xE9\x1E\x63
\xE9\x1E\x63
\xE9\x1E\x63
\xE9\x1E\x63
\xE9\x1E\x63
\xE9\x1E\x63
\xE9\x1E\x63
\xE9\x1E\x63
\xE9\x1E\x63
\xE9\x1E\x63
\xE9\x1E\x63
\xE9\x1E\x63
\xE9\x1E\x63
\xE9\x1E\x63
\xff\xff\xff,\0\0\0\0\xc8\0\x46\0\0\x04`,

	// Purple 500 #9C27B0
	`GIF89a\xc8\0\x46\0\x83\0\0
\x9C\x27\xB0
\x9C\x27\xB0
\x9C\x27\xB0
\x9C\x27\xB0
\x9C\x27\xB0
\x9C\x27\xB0
\x9C\x27\xB0
\x9C\x27\xB0
\x9C\x27\xB0
\x9C\x27\xB0
\x9C\x27\xB0
\x9C\x27\xB0
\x9C\x27\xB0
\x9C\x27\xB0
\x9C\x27\xB0
\xff\xff\xff,\0\0\0\0\xc8\0\x46\0\0\x04`,

	// Deep Purple 500 #673AB7
	`GIF89a\xc8\0\x46\0\x83\0\0
\x67\x3A\xB7
\x67\x3A\xB7
\x67\x3A\xB7
\x67\x3A\xB7
\x67\x3A\xB7
\x67\x3A\xB7
\x67\x3A\xB7
\x67\x3A\xB7
\x67\x3A\xB7
\x67\x3A\xB7
\x67\x3A\xB7
\x67\x3A\xB7
\x67\x3A\xB7
\x67\x3A\xB7
\x67\x3A\xB7
\xff\xff\xff,\0\0\0\0\xc8\0\x46\0\0\x04`,

	// Indigo 500 #3F51B5
	`GIF89a\xc8\0\x46\0\x83\0\0
\x3F\x51\xB5
\x3F\x51\xB5
\x3F\x51\xB5
\x3F\x51\xB5
\x3F\x51\xB5
\x3F\x51\xB5
\x3F\x51\xB5
\x3F\x51\xB5
\x3F\x51\xB5
\x3F\x51\xB5
\x3F\x51\xB5
\x3F\x51\xB5
\x3F\x51\xB5
\x3F\x51\xB5
\x3F\x51\xB5
\xff\xff\xff,\0\0\0\0\xc8\0\x46\0\0\x04`,

	// Blue 500 #2196F3
	`GIF89a\xc8\0\x46\0\x83\0\0
\x21\x96\xF3
\x21\x96\xF3
\x21\x96\xF3
\x21\x96\xF3
\x21\x96\xF3
\x21\x96\xF3
\x21\x96\xF3
\x21\x96\xF3
\x21\x96\xF3
\x21\x96\xF3
\x21\x96\xF3
\x21\x96\xF3
\x21\x96\xF3
\x21\x96\xF3
\x21\x96\xF3
\xff\xff\xff,\0\0\0\0\xc8\0\x46\0\0\x04`,

	// Light Blue 500 #03A9F4
	`GIF89a\xc8\0\x46\0\x83\0\0
\x03\xA9\xF4
\x03\xA9\xF4
\x03\xA9\xF4
\x03\xA9\xF4
\x03\xA9\xF4
\x03\xA9\xF4
\x03\xA9\xF4
\x03\xA9\xF4
\x03\xA9\xF4
\x03\xA9\xF4
\x03\xA9\xF4
\x03\xA9\xF4
\x03\xA9\xF4
\x03\xA9\xF4
\x03\xA9\xF4
\xff\xff\xff,\0\0\0\0\xc8\0\x46\0\0\x04`,

	// Cyan 500 #00BCD4
	`GIF89a\xc8\0\x46\0\x83\0\0
\x00\xBC\xD4
\x00\xBC\xD4
\x00\xBC\xD4
\x00\xBC\xD4
\x00\xBC\xD4
\x00\xBC\xD4
\x00\xBC\xD4
\x00\xBC\xD4
\x00\xBC\xD4
\x00\xBC\xD4
\x00\xBC\xD4
\x00\xBC\xD4
\x00\xBC\xD4
\x00\xBC\xD4
\x00\xBC\xD4
\xff\xff\xff,\0\0\0\0\xc8\0\x46\0\0\x04`,

	// Teal 500 #009688
	`GIF89a\xc8\0\x46\0\x83\0\0
\x00\x96\x88
\x00\x96\x88
\x00\x96\x88
\x00\x96\x88
\x00\x96\x88
\x00\x96\x88
\x00\x96\x88
\x00\x96\x88
\x00\x96\x88
\x00\x96\x88
\x00\x96\x88
\x00\x96\x88
\x00\x96\x88
\x00\x96\x88
\x00\x96\x88
\xff\xff\xff,\0\0\0\0\xc8\0\x46\0\0\x04`,

	// Green 500 #4CAF50
	`GIF89a\xc8\0\x46\0\x83\0\0
\x4C\xAF\x50
\x4C\xAF\x50
\x4C\xAF\x50
\x4C\xAF\x50
\x4C\xAF\x50
\x4C\xAF\x50
\x4C\xAF\x50
\x4C\xAF\x50
\x4C\xAF\x50
\x4C\xAF\x50
\x4C\xAF\x50
\x4C\xAF\x50
\x4C\xAF\x50
\x4C\xAF\x50
\x4C\xAF\x50
\xff\xff\xff,\0\0\0\0\xc8\0\x46\0\0\x04`,

	// Light Green 500 #8BC34A
	`GIF89a\xc8\0\x46\0\x83\0\0
\x8B\xC3\x4A
\x8B\xC3\x4A
\x8B\xC3\x4A
\x8B\xC3\x4A
\x8B\xC3\x4A
\x8B\xC3\x4A
\x8B\xC3\x4A
\x8B\xC3\x4A
\x8B\xC3\x4A
\x8B\xC3\x4A
\x8B\xC3\x4A
\x8B\xC3\x4A
\x8B\xC3\x4A
\x8B\xC3\x4A
\x8B\xC3\x4A
\xff\xff\xff,\0\0\0\0\xc8\0\x46\0\0\x04`,

	// Lime 500 #CDDC39
	`GIF89a\xc8\0\x46\0\x83\0\0
\xCD\xDC\x39
\xCD\xDC\x39
\xCD\xDC\x39
\xCD\xDC\x39
\xCD\xDC\x39
\xCD\xDC\x39
\xCD\xDC\x39
\xCD\xDC\x39
\xCD\xDC\x39
\xCD\xDC\x39
\xCD\xDC\x39
\xCD\xDC\x39
\xCD\xDC\x39
\xCD\xDC\x39
\xCD\xDC\x39
\xff\xff\xff,\0\0\0\0\xc8\0\x46\0\0\x04`,

	// Yellow 500 #FFEB3B
	`GIF89a\xc8\0\x46\0\x83\0\0
\xFF\xEB\x3B
\xFF\xEB\x3B
\xFF\xEB\x3B
\xFF\xEB\x3B
\xFF\xEB\x3B
\xFF\xEB\x3B
\xFF\xEB\x3B
\xFF\xEB\x3B
\xFF\xEB\x3B
\xFF\xEB\x3B
\xFF\xEB\x3B
\xFF\xEB\x3B
\xFF\xEB\x3B
\xFF\xEB\x3B
\xFF\xEB\x3B
\xff\xff\xff,\0\0\0\0\xc8\0\x46\0\0\x04`,

	// Amber 500 #FFC107
	`GIF89a\xc8\0\x46\0\x83\0\0
\xFF\xC1\x07
\xFF\xC1\x07
\xFF\xC1\x07
\xFF\xC1\x07
\xFF\xC1\x07
\xFF\xC1\x07
\xFF\xC1\x07
\xFF\xC1\x07
\xFF\xC1\x07
\xFF\xC1\x07
\xFF\xC1\x07
\xFF\xC1\x07
\xFF\xC1\x07
\xFF\xC1\x07
\xFF\xC1\x07
\xff\xff\xff,\0\0\0\0\xc8\0\x46\0\0\x04`,

	// Orange 500 #FF9800
	`GIF89a\xc8\0\x46\0\x83\0\0
\xFF\x98\x00
\xFF\x98\x00
\xFF\x98\x00
\xFF\x98\x00
\xFF\x98\x00
\xFF\x98\x00
\xFF\x98\x00
\xFF\x98\x00
\xFF\x98\x00
\xFF\x98\x00
\xFF\x98\x00
\xFF\x98\x00
\xFF\x98\x00
\xFF\x98\x00
\xFF\x98\x00
\xff\xff\xff,\0\0\0\0\xc8\0\x46\0\0\x04`,

	// Deep Orange 500 #FF5722
	`GIF89a\xc8\0\x46\0\x83\0\0
\xFF\x57\x22
\xFF\x57\x22
\xFF\x57\x22
\xFF\x57\x22
\xFF\x57\x22
\xFF\x57\x22
\xFF\x57\x22
\xFF\x57\x22
\xFF\x57\x22
\xFF\x57\x22
\xFF\x57\x22
\xFF\x57\x22
\xFF\x57\x22
\xFF\x57\x22
\xFF\x57\x22
\xff\xff\xff,\0\0\0\0\xc8\0\x46\0\0\x04`,

	// Brown 500 #795548
	`GIF89a\xc8\0\x46\0\x83\0\0
\x79\x55\x48
\x79\x55\x48
\x79\x55\x48
\x79\x55\x48
\x79\x55\x48
\x79\x55\x48
\x79\x55\x48
\x79\x55\x48
\x79\x55\x48
\x79\x55\x48
\x79\x55\x48
\x79\x55\x48
\x79\x55\x48
\x79\x55\x48
\x79\x55\x48
\xff\xff\xff,\0\0\0\0\xc8\0\x46\0\0\x04`
] as const;


src/lib/Captcha.ts
import { randomBytes } from 'node:crypto';
import { colors } from './Colors';
import { allLetters } from './Font';
import { SW } from './Sw';

/**
 * The captcha generator
 */
export class Captcha {
	private readonly numberOfDots: number;
	private readonly gifSize = 17646;
	private readonly letters = 'abcdefghijklmnopqrstuvwxyz';
	private readonly blurImg: boolean;
	private readonly filterImg: boolean;

	public constructor({ numberOfDots = 100, blur = false, filter = false }: Options = {}) {
		this.numberOfDots = numberOfDots;
		this.blurImg = blur;
		this.filterImg = filter;
	}

	/**
	 *
	 * @param size The number of letters
	 * @param color The color of the text
	 * @returns The gif buffer and the token
	 */
	public generate(size = 5, color?: Color) {
		const buffer = Buffer.alloc(this.gifSize);
		const { background, token } = this.makeCaptcha(size);
		this.makeGif(background, buffer, color);

		return {
			buffer,
			token
		};
	}

	/**
	 *
	 * @param min min size
	 * @param max max size
	 * @returns A number between min and max
	 */
	private random(min: number, max: number) {
		return Math.floor(Math.random() * (max - min)) + min;
	}

	/**
	 *
	 * @param n The index of the letter
	 * @param pos The position of the letter
	 * @param background The background buffer
	 * @param swr The sw buffer
	 * @param s1 The s1 value
	 * @param s2 The s2 value
	 * @returns The new position of the letter
	 */
	private letter(n: number, pos: number, background: Buffer, swr: Buffer, s1: number, s2: number) {
		const l = background.length;
		const t = allLetters[n];
		let r = 200 * 16 + pos;
		let i = r;
		let sk1 = s1 + pos;
		let sk2 = s2 + pos;
		let mpos = pos;
		let row = 0;

		for (let j = 0, k = t.length; j < k; j++) {
			const p = t[j];
			if (p === -101) continue;

			if (p < 0) {
				if (p === -100) {
					r += 200;
					i = r;
					sk1 = s1 + pos;
					row++;
					continue;
				}
				i += -p;
				continue;
			}

			if (sk1 >= 200) sk1 %= 200;
			const skew = Math.floor(SW[sk1] / 16);
			sk1 += (swr[pos + i - r] & 0x1) + 1;

			if (sk2 >= 200) sk2 %= 200;
			const skewh = Math.floor(SW[sk2] / 70);
			sk2 += swr[row] & 0x1;

			const x = i + skew * 200 + skewh;
			mpos = Math.max(mpos, pos + i - r);

			if (x - l < 70 * 200) background[x] = p << 4;
			i++;
		}

		return mpos;
	}

	/**
	 *
	 * @param background The background buffer
	 * @param swr The sw buffer
	 * @param s1 The s1 value
	 */
	private line(background: Buffer, swr: Buffer, s1: number) {
		for (let x = 0, sk1 = s1; x < 199; x++) {
			if (sk1 >= 200) sk1 %= 200;
			const skew = Math.floor(SW[sk1] / 16);
			sk1 += (swr[x] & 0x3) + 1;
			const i = 200 * (45 + skew) + x;
			background[i + 0] = 0;
			background[i + 1] = 0;
			background[i + 200] = 0;
			background[i + 201] = 0;
		}
	}

	/**
	 *
	 * @param background The background buffer
	 * @param dr The dr buffer
	 */
	private dots(background: Buffer, dr: Buffer) {
		for (let n = 0; n < this.numberOfDots; n++) {
			const v = dr.readUInt32BE(n);
			const i = v % (200 * 67);

			background[i + 0] = 0xff;
			background[i + 1] = 0xff;
			background[i + 2] = 0xff;
			background[i + 200] = 0xff;
			background[i + 201] = 0xff;
			background[i + 202] = 0xff;
		}
	}

	/**
	 *
	 * @param background The background buffer
	 */
	private blur(background: Buffer) {
		for (let i = 0, y = 0; y < 68; y++) {
			for (let x = 0; x < 198; x++) {
				const c11 = background[i + 0];
				const c12 = background[i + 1];
				const c21 = background[i + 200];
				const c22 = background[i + 201];
				background[i++] = Math.floor((c11 + c12 + c21 + c22) / 4);
			}
		}
	}

	/**
	 *
	 * @param background The background buffer
	 */
	private filter(background: Buffer) {
		const om = Buffer.alloc(70 * 200).fill(0xff);
		let i = 0;
		let o = 0;

		for (let y = 0; y < 70; y++) {
			for (let x = 4; x < 200 - 4; x++) {
				if (background[i] > 0xf0 && background[i + 1] < 0xf0) {
					om[o] = 0;
					om[o + 1] = 0;
				} else if (background[i] < 0xf0 && background[i + 1] > 0xf0) {
					om[o] = 0;
					om[o + 1] = 0;
				}

				i++;
				o++;
			}
		}

		om.copy(background);
	}

	/**
	 *
	 * @param size The number of letters
	 * @returns The background buffer and the token
	 */
	private makeCaptcha(size: number) {
		const rb = randomBytes(size + 200 + 100 * 4 + 1 + 1);
		const tb = Buffer.alloc(size);
		const swr = Buffer.alloc(200);
		const dr = Buffer.alloc(100 * 4);

		rb.copy(tb, 0, 0, size);
		rb.copy(swr, 0, size, size + 200);
		rb.copy(dr, 0, size + 200, size + 200 + 100 * 4);
		let s1 = rb.readUInt8(size + 200 + 100 * 4);
		let s2 = rb.readUInt8(size + 200 + 100 * 4 + 1);

		const background = Buffer.alloc(200 * 70).fill(0xff);

		s1 &= 0x7f;
		s2 &= 0x3f;

		let p = 30;

		for (let n = 0; n < size; n++) {
			tb[n] %= 25;
			p = this.letter(tb[n], p, background, swr, s1, s2);
			tb[n] = this.letters.charCodeAt(tb[n]);
		}

		this.line(background, swr, s1);
		this.dots(background, dr);
		if (this.blurImg) this.blur(background);
		if (this.filterImg) this.filter(background);

		return { background, token: tb.toString() };
	}

	/**
	 *
	 * @param background The background buffer
	 * @param gif The gif buffer
	 * @param color The color of the text
	 *
	 */
	private makeGif(background: Buffer, gif: Buffer, color?: Color) {
		const r = color ?? (this.random(0, colors.length) as Color);
		gif.fill(colors[r].replace(/\n/g, ''), 0, 13 + 48 + 10 + 1, 'ascii');

		let i = 0;
		let p = 13 + 48 + 10 + 1;
		for (let y = 0; y < 70; y++) {
			gif[p++] = 250; // Data length 5*50=250
			for (let x = 0; x < 50; x++) {
				const a = background[i + 0] >> 4;
				const b = background[i + 1] >> 4;
				const c = background[i + 2] >> 4;
				const d = background[i + 3] >> 4;

				gif[p + 0] = 16 | (a << 5); // bbb10000
				gif[p + 1] = (a >> 3) | 64 | (b << 7); // b10000xb
				gif[p + 2] = b >> 1; // 0000xbbb
				gif[p + 3] = 1 | (c << 1); // 00xbbbb1
				gif[p + 4] = 4 | (d << 3); // xbbbb100
				i += 4;
				p += 5;
			}
		}

		gif.fill('\x01\x11\x00;', this.gifSize - 4);
	}
}

/**
 * The options for the captcha constructor
 */
export interface Options {
	numberOfDots?: number;
	blur?: boolean;
	filter?: boolean;
}

export type Color = 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 | 16 | 17;
